### PR TITLE
Implement RegExp `v` flag with set notation + properties of strings

### PR DESCRIPTION
**Pull request by msaboff**
#### 270c824459cec4d19dab347a8db1526e0be50737
<pre>
Implement RegExp `v` flag with set notation + properties of strings
<a href="https://bugs.webkit.org/show_bug.cgi?id=241593">https://bugs.webkit.org/show_bug.cgi?id=241593</a>
rdar://100337109

Reviewed by Yusuke Suzuki.

This change implements the TC39 stage 3 proposal RegExp v flag with set notation + properties of strings,
<a href="https://github.com/tc39/proposal-regexp-v-flag.">https://github.com/tc39/proposal-regexp-v-flag.</a>  It adds a new &quot;unicodeSets&quot; compile mode for the Yarr engine.
Like the prior Unicode Yarr features, this change is driven by Unicode Database Files (UCD).
This change includes two such new files, JavaScriptCore/ucd/{emoji-sequences.txt &amp; emoji-zwj-sequences.txt}.

The newly added properties include lists of strings.  These strings are processed via the character class syntax
through.  When it comes to matching however, there is some desuguraing that turns such a property of strings into
a list of alternations.  For example, say a property has strings str1...strN plus a traditional character class,
single-character-class, we create the pattern equivalent of:
     (?:str1|str2|...|strN|[single-character-class])
Per the spec, longer strings appear earlier in the alternation, and before the traditional character class.
This allows for searching for longer properties in a property list where substrings of other strings are included
in that list.

There are new set of combining operators allowed in the class set character classes.  Two character class elements
that appear adjacent to each other implicitly have the Union combining operations.  There is also an Intersection
operation with the &amp;&amp; operator and a Subtraction operation with the || operator.

There is new ClassSet parsing that follows new &quot;cleaner&quot; rules that traditional character classes.
The prior ccharacter class constructor and delegates are mostly unchanged, except for the compile mode
now being switched on an enum instead of a bool.

Added check that both &apos;u&apos; and &apos;v&apos; flags don&apos;t appear in the same RegExp.

Added unicodeSets getter watchpoint to the m_regExpPrimordialPropertiesWatchpointSet.

* JSTests/es6/Proxy_internal_get_calls_RegExp.prototype.flags.js:
* JSTests/stress/regexp-vflag-property-of-strings.js: Added.
(arrayToString):
(objectToString):
(dumpValue):
(compareArray):
(compareGroups):
(testRegExp):
(testRegExpSyntaxError):
* JSTests/stress/static-getter-in-names.js:
* JSTests/test262/config.yaml:
* LayoutTests/js/Object-getOwnPropertyNames-expected.txt:
* LayoutTests/js/script-tests/Object-getOwnPropertyNames.js:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/RegExpPrototype.js:
(linkTimeConstant.hasObservableSideEffectsForRegExpMatch):
(linkTimeConstant.hasObservableSideEffectsForRegExpSplit):
(overriddenName.string_appeared_here.split):
* Source/JavaScriptCore/builtins/StringPrototype.js:
(linkTimeConstant.hasObservableSideEffectsForStringReplace):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::addStringReplacePrimordialChecks):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
(JSC::DFG::StrengthReductionPhase::handleNode):
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::regExpProtoUnicodeSetsGetter const):
* Source/JavaScriptCore/runtime/RegExp.h:
* Source/JavaScriptCore/runtime/RegExpCache.h:
* Source/JavaScriptCore/runtime/RegExpObject.cpp:
(JSC::RegExpObject::matchGlobal):
* Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
(JSC::RegExpPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/ucd/emoji-sequences.txt: Added.
* Source/JavaScriptCore/ucd/emoji-zwj-sequences.txt: Added.
* Source/JavaScriptCore/yarr/Yarr.h:
* Source/JavaScriptCore/yarr/YarrErrorCode.cpp:
(JSC::Yarr::errorMessage):
(JSC::Yarr::errorToThrow):
* Source/JavaScriptCore/yarr/YarrErrorCode.h:
* Source/JavaScriptCore/yarr/YarrFlags.h:
* Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
(JSC::Yarr::ByteTermDumper::ByteTermDumper):
(JSC::Yarr::ByteTermDumper::unicode const):
(JSC::Yarr::ByteTermDumper::unicodeSets const):
(JSC::Yarr::ByteTermDumper::eitherUnicode const):
(JSC::Yarr::Interpreter::tryConsumeBackReference):
(JSC::Yarr::Interpreter::matchCharacterClass):
(JSC::Yarr::Interpreter::backtrackCharacterClass):
(JSC::Yarr::Interpreter::matchDisjunction):
(JSC::Yarr::Interpreter::Interpreter):
(JSC::Yarr::Interpreter::isLegacyCompilation const):
(JSC::Yarr::Interpreter::isUnicodeCompilation const):
(JSC::Yarr::Interpreter::isUnicodeSetsCompilation const):
(JSC::Yarr::Interpreter::isEitherUnicodeCompilation const):
(JSC::Yarr::ByteTermDumper::dumpTerm):
(JSC::Yarr::ByteTermDumper::unicode): Deleted.
* Source/JavaScriptCore/yarr/YarrInterpreter.h:
(JSC::Yarr::BytecodePattern::BytecodePattern):
(JSC::Yarr::BytecodePattern::compileMode const):
(JSC::Yarr::BytecodePattern::unicodeSets const):
(JSC::Yarr::BytecodePattern::eitherUnicode const):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
* Source/JavaScriptCore/yarr/YarrParser.h:
(JSC::Yarr::Parser::CharacterClassParserDelegate::CharacterClassParserDelegate):
(JSC::Yarr::Parser::ClassSetParserDelegate::ClassSetParserDelegate):
(JSC::Yarr::Parser::ClassSetParserDelegate::begin):
(JSC::Yarr::Parser::ClassSetParserDelegate::nestedClassBegin):
(JSC::Yarr::Parser::ClassSetParserDelegate::doneAfterCharacterClassEnd):
(JSC::Yarr::Parser::ClassSetParserDelegate::setUnionOp):
(JSC::Yarr::Parser::ClassSetParserDelegate::setSubtractOp):
(JSC::Yarr::Parser::ClassSetParserDelegate::setIntersectionOp):
(JSC::Yarr::Parser::ClassSetParserDelegate::flushCachedCharacterIfNeeded):
(JSC::Yarr::Parser::ClassSetParserDelegate::atomPatternCharacter):
(JSC::Yarr::Parser::ClassSetParserDelegate::atomBuiltInCharacterClass):
(JSC::Yarr::Parser::ClassSetParserDelegate::end):
(JSC::Yarr::Parser::ClassSetParserDelegate::error):
(JSC::Yarr::Parser::ClassSetParserDelegate::assertionWordBoundary):
(JSC::Yarr::Parser::ClassSetParserDelegate::atomBackReference):
(JSC::Yarr::Parser::ClassSetParserDelegate::atomNamedBackReference):
(JSC::Yarr::Parser::ClassSetParserDelegate::atomNamedForwardReference):
(JSC::Yarr::Parser::ClassStringDisjunctionParserDelegate::ClassStringDisjunctionParserDelegate):
(JSC::Yarr::Parser::ClassStringDisjunctionParserDelegate::atomPatternCharacter):
(JSC::Yarr::Parser::ClassStringDisjunctionParserDelegate::newAlternative):
(JSC::Yarr::Parser::ClassStringDisjunctionParserDelegate::end):
(JSC::Yarr::Parser::ClassStringDisjunctionParserDelegate::assertionWordBoundary):
(JSC::Yarr::Parser::ClassStringDisjunctionParserDelegate::atomBackReference):
(JSC::Yarr::Parser::ClassStringDisjunctionParserDelegate::atomNamedBackReference):
(JSC::Yarr::Parser::ClassStringDisjunctionParserDelegate::atomNamedForwardReference):
(JSC::Yarr::Parser::ClassStringDisjunctionParserDelegate::atomBuiltInCharacterClass):
(JSC::Yarr::Parser::Parser):
(JSC::Yarr::Parser::isIdentityEscapeAnError):
(JSC::Yarr::Parser::parseEscape):
(JSC::Yarr::Parser::consumePossibleSurrogatePair):
(JSC::Yarr::Parser::parseAtomEscape):
(JSC::Yarr::Parser::parseCharacterClassEscape):
(JSC::Yarr::Parser::parseClassSetEscape):
(JSC::Yarr::Parser::parseClassStringDisjunctionEscape):
(JSC::Yarr::Parser::parseCharacterClass):
(JSC::Yarr::Parser::parseClassSet):
(JSC::Yarr::Parser::parseClassStringDisjunction):
(JSC::Yarr::Parser::parseParenthesesEnd):
(JSC::Yarr::Parser::parseTokens):
(JSC::Yarr::Parser::handleIllegalReferences):
(JSC::Yarr::Parser::tryConsumeUnicodeEscape):
(JSC::Yarr::Parser::tryConsumeUnicodePropertyExpression):
(JSC::Yarr::Parser::isLegacyCompilation const):
(JSC::Yarr::Parser::isUnicodeCompilation const):
(JSC::Yarr::Parser::isUnicodeSetsCompilation const):
(JSC::Yarr::Parser::isEitherUnicodeCompilation const):
(JSC::Yarr::compileMode):
(JSC::Yarr::parse):
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::CharacterClassConstructor::CharacterClassConstructor):
(JSC::Yarr::CharacterClassConstructor::reset):
(JSC::Yarr::CharacterClassConstructor::combiningSetOp):
(JSC::Yarr::CharacterClassConstructor::append):
(JSC::Yarr::CharacterClassConstructor::appendInverted):
(JSC::Yarr::CharacterClassConstructor::putRange):
(JSC::Yarr::CharacterClassConstructor::atomClassStringDisjunction):
(JSC::Yarr::CharacterClassConstructor::performSetOpWith):
(JSC::Yarr::CharacterClassConstructor::performSetOpWithStrings):
(JSC::Yarr::CharacterClassConstructor::performSetOpWithMatches):
(JSC::Yarr::CharacterClassConstructor::hasInverteStrings):
(JSC::Yarr::CharacterClassConstructor::compareUTF32Strings):
(JSC::Yarr::CharacterClassConstructor::sort):
(JSC::Yarr::CharacterClassConstructor::charClass):
(JSC::Yarr::CharacterClassConstructor::mergeRangesFrom):
(JSC::Yarr::CharacterClassConstructor::unionStrings):
(JSC::Yarr::CharacterClassConstructor::intersectionStrings):
(JSC::Yarr::CharacterClassConstructor::subtractionStrings):
(JSC::Yarr::CharacterClassConstructor::asciiOpSorted):
(JSC::Yarr::CharacterClassConstructor::unicodeOpSorted):
(JSC::Yarr::YarrPatternConstructor::YarrPatternConstructor):
(JSC::Yarr::YarrPatternConstructor::resetForReparsing):
(JSC::Yarr::YarrPatternConstructor::atomPatternCharacter):
(JSC::Yarr::YarrPatternConstructor::atomBuiltInCharacterClass):
(JSC::Yarr::YarrPatternConstructor::atomCharacterClassAtom):
(JSC::Yarr::YarrPatternConstructor::atomCharacterClassRange):
(JSC::Yarr::YarrPatternConstructor::atomCharacterClassBuiltIn):
(JSC::Yarr::YarrPatternConstructor::atomClassStringDisjunction):
(JSC::Yarr::YarrPatternConstructor::atomCharacterClassSetOp):
(JSC::Yarr::YarrPatternConstructor::atomCharacterClassPushNested):
(JSC::Yarr::YarrPatternConstructor::atomCharacterClassPopNested):
(JSC::Yarr::YarrPatternConstructor::atomCharacterClassEnd):
(JSC::Yarr::YarrPattern::compile):
(JSC::Yarr::PatternTerm::dump):
(JSC::Yarr::YarrPattern::dumpPatternString):
(JSC::Yarr::YarrPattern::dumpPattern):
* Source/JavaScriptCore/yarr/YarrPattern.h:
(JSC::Yarr::CharacterClass::CharacterClass):
(JSC::Yarr::CharacterClass::hasNonBMPCharacters const):
(JSC::Yarr::CharacterClass::hasOneCharacterSize const):
(JSC::Yarr::CharacterClass::hasOnlyNonBMPCharacters const):
(JSC::Yarr::CharacterClass::hasStrings const):
(JSC::Yarr::CharacterClass::hasSingleCharacters const):
(JSC::Yarr::ClassSet::ClassSet):
(JSC::Yarr::YarrPattern::unicodeSets const):
(JSC::Yarr::YarrPattern::eitherUnicode const):
(JSC::Yarr::YarrPattern::compileMode const):
(JSC::Yarr::CharacterClass::hasNonBMPCharacters): Deleted.
(JSC::Yarr::CharacterClass::hasOneCharacterSize): Deleted.
(JSC::Yarr::CharacterClass::hasOnlyNonBMPCharacters): Deleted.
* Source/JavaScriptCore/yarr/YarrSyntaxChecker.cpp:
(JSC::Yarr::SyntaxChecker::atomClassStringDisjunction):
(JSC::Yarr::SyntaxChecker::atomCharacterClassSetOp):
(JSC::Yarr::SyntaxChecker::atomCharacterClassPushNested):
(JSC::Yarr::SyntaxChecker::atomCharacterClassPopNested):
(JSC::Yarr::checkSyntax):
* Source/JavaScriptCore/yarr/YarrUnicodeProperties.cpp:
(JSC::Yarr::unicodeMatchProperty):
(JSC::Yarr::createUnicodeCharacterClassFor):
(JSC::Yarr::characterClassMayContainStrings):
* Source/JavaScriptCore/yarr/YarrUnicodeProperties.h:
* Source/JavaScriptCore/yarr/generateYarrUnicodePropertyTables.py:
(PropertyData.__init__):
(PropertyData.addMatchString):
(PropertyData.stringsCompare):
(PropertyData):
(PropertyData.sortStrings):
(PropertyData.dumpMatchData):
(PropertyData.convertStringToCppFormat):
(PropertyData.dump):
(PropertyData.dumpAll):
(PropertyData.dumpMayContainStringFunc):
(BinaryProperty.dump):
(SequenceProperty):
(SequenceProperty.__init__):
(SequenceProperty.parsePropertyFile):
(SequenceProperty.dump):
* Source/WebCore/contentextensions/URLFilterParser.cpp:
(WebCore::ContentExtensions::PatternParser::atomClassStringDisjunction):
(WebCore::ContentExtensions::PatternParser::atomCharacterClassSetOp):
(WebCore::ContentExtensions::PatternParser::atomCharacterClassPushNested):
(WebCore::ContentExtensions::PatternParser::atomCharacterClassPopNested):
(WebCore::ContentExtensions::URLFilterParser::addPattern):

Canonical link: <a href="https://commits.webkit.org/261188@main">https://commits.webkit.org/261188@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/033aaa0704ab6118341a4972f81be5e98cbb4dc9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/110856 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/19942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/43404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/2210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/114817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/21367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/11078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/87/builds/2210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116608 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/21367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/43404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/103223 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/21367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/43404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/103223 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/99487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/12536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/43404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/103223 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/10635 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/13093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/11078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/100589 "Built successfully") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/18494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/43404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/31373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/15038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/108620 "Built successfully") | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4231 "Failed to push commit to Webkit repository") | | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/26777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") |

### CODE REVIEW

File: Source/JavaScriptCore/builtins/StringPrototype.js [hunk 1]
```diff
@@ -234,6 +234,10 @@ function hasObservableSideEffectsForStringReplace(regexp, replacer)
     if (regexpUnicode !== @regExpProtoUnicodeGetter)
         return true;
 
+    var regexpUnicode = @tryGetById(regexp, "unicodeSets");
+    if (regexpUnicode !== @regExpProtoUnicodeGetter)
+        return true;
+
```
Comment (Constellation) [inline Source/JavaScriptCore/builtins/StringPrototype.js]: You need to add optimization for `unicodeSets`. See JSGlobalObject.cpp etc. This is comparing with `@regExpProtoUnicodeGetter`, which is `unicode` getter.

File: Source/JavaScriptCore/builtins/RegExpPrototype.js [hunk 2]
```diff
@@ -460,7 +464,10 @@ function hasObservableSideEffectsForRegExpSplit(regexp)
     var regexpUnicode = @tryGetById(regexp, "unicode");
     if (regexpUnicode !== @regExpProtoUnicodeGetter)
         return true;
-    
+    var regexpUnicode = @tryGetById(regexp, "unicodeSets");
+    if (regexpUnicode !== @regExpProtoUnicodeGetter)
+        return true;
```
Comment (Constellation) [inline Source/JavaScriptCore/builtins/RegExpPrototype.js]: Ditto.

File: Source/JavaScriptCore/builtins/RegExpPrototype.js [hunk 3]
```diff
@@ -95,6 +95,10 @@ function hasObservableSideEffectsForRegExpMatch(regexp)
     if (regexpUnicode !== @regExpProtoUnicodeGetter)
         return true;
 
+    var regexpUnicode = @tryGetById(regexp, "unicodeSets");
+    if (regexpUnicode !== @regExpProtoUnicodeGetter)
+        return true;
```
Comment (Constellation) [inline Source/JavaScriptCore/builtins/RegExpPrototype.js]: Ditto. This is `unicode` getter.

File: Source/JavaScriptCore/builtins/StringPrototype.js (hunk 1 above)
Comment (Constellation) [reply Source/JavaScriptCore/builtins/StringPrototype.js]: Also need to add optimization done for `unicode` getter etc. Like, DFGAbstractInterpreterInlines.h etc. for example.

Comment (Constellation) [reply Source/JavaScriptCore/builtins/StringPrototype.js]: See `DFGFixupPhase.cpp` and `DFGAbstractInterpreterInlines.h` too. Let's grep with `propertyNames->global` for example.

**Review by rkirsling — CHANGES_REQUESTED** [review]
Please be sure to update the test262 results in expectations.yaml; in addition to the 7(x2) outstanding failures introduced by #10355 (displayed [here](https://github.com/WebKit/WebKit/pull/10851#issuecomment-1449537623)), this patch appears to introduce an additional 2(x2) failures:
```yaml
test/built-ins/RegExp/prototype/unicodeSets/uv-flags-constructor.js:
  default: 'Test262Error: Expected a SyntaxError to be thrown but no exception was thrown at all'
  strict mode: 'Test262Error: Expected a SyntaxError to be thrown but no exception was thrown at all'
test/built-ins/RegExp/prototype/unicodeSets/uv-flags.js:
  default: 'Test262: This statement should not be evaluated.'
  strict mode: 'Test262: This statement should not be evaluated.'
```

File: Source/JavaScriptCore/builtins/StringPrototype.js (hunk 1 above)
Comment (msaboff) [reply Source/JavaScriptCore/builtins/StringPrototype.js]: I fixed these.

File: Source/JavaScriptCore/builtins/RegExpPrototype.js (hunk 2 above)
Comment (msaboff) [reply Source/JavaScriptCore/builtins/RegExpPrototype.js]: Fixed.

File: Source/JavaScriptCore/builtins/RegExpPrototype.js (hunk 3 above)
Comment (msaboff) [reply Source/JavaScriptCore/builtins/RegExpPrototype.js]: Fixed.

**Comment by msaboff** [comment]
> Please be sure to update the test262 results in expectations.yaml; in addition to the 7(x2) outstanding failures introduced by #10355 (displayed [here](https://github.com/WebKit/WebKit/pull/10851#issuecomment-1449537623)), this patch appears to introduce an additional 2(x2) failures:
> 
> ```yaml
> test/built-ins/RegExp/prototype/unicodeSets/uv-flags-constructor.js:
>   default: 'Test262Error: Expected a SyntaxError to be thrown but no exception was thrown at all'
>   strict mode: 'Test262Error: Expected a SyntaxError to be thrown but no exception was thrown at all'
> test/built-ins/RegExp/prototype/unicodeSets/uv-flags.js:
>   default: 'Test262: This statement should not be evaluated.'
>   strict mode: 'Test262: This statement should not be evaluated.'
> ```

@rkirsling I'll update test262 results in a follow on PR.

File: Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h [hunk 4]
```diff
@@ -3555,6 +3555,9 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
                     if (attemptToFold(m_vm.propertyNames->unicode.impl(), globalObject->regExpProtoUnicodeGetter()))
                         break;
 
+                    if (attemptToFold(m_vm.propertyNames->unicode.impl(), globalObject->regExpProtoUnicodeSetsGetter()))
```
Comment (Constellation) [inline Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h]: Not `unicode`.

Comment (msaboff) [reply Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h]: Copy paste error - Fixed.

**Comment by msaboff** [comment]
@rkirsling: I fixed a minor issue where I wasn't checking for both 'v' and 'u' for the same RegExp.  After that change, all test262 V Flag / unicodeSet tests pass, so no update needed beside the included config.yaml.

**Comment by msaboff** [comment]
> Please be sure to update the test262 results in expectations.yaml; in addition to the 7(x2) outstanding failures introduced by #10355 (displayed [here](https://github.com/WebKit/WebKit/pull/10851#issuecomment-1449537623)), this patch appears to introduce an additional 2(x2) failures:
> 
> ```yaml
> test/built-ins/RegExp/prototype/unicodeSets/uv-flags-constructor.js:
>   default: 'Test262Error: Expected a SyntaxError to be thrown but no exception was thrown at all'
>   strict mode: 'Test262Error: Expected a SyntaxError to be thrown but no exception was thrown at all'
> test/built-ins/RegExp/prototype/unicodeSets/uv-flags.js:
>   default: 'Test262: This statement should not be evaluated.'
>   strict mode: 'Test262: This statement should not be evaluated.'
> ```

Fixed these failures in the latest version.

File: Source/JavaScriptCore/runtime/JSGlobalObject.cpp [hunk 5]
```diff
@@ -1416,6 +1416,9 @@ capitalName ## Constructor* lowerName ## Constructor = featureFlag ? capitalName
     GetterSetter* regExpProtoUnicodeGetter = getGetterById(this, m_regExpPrototype.get(), vm.propertyNames->unicode);
     catchScope.assertNoException();
     m_linkTimeConstants[static_cast<unsigned>(LinkTimeConstant::regExpProtoUnicodeGetter)].set(vm, this, regExpProtoUnicodeGetter);
+    GetterSetter* regExpProtoUnicodeSetsGetter = getGetterById(this, m_regExpPrototype.get(), vm.propertyNames->unicodeSets);
+    catchScope.assertNoException();
+    m_linkTimeConstants[static_cast<unsigned>(LinkTimeConstant::regExpProtoUnicodeSetsGetter)].set(vm, this, regExpProtoUnicodeSetsGetter);
```
Comment (Constellation) [inline Source/JavaScriptCore/runtime/JSGlobalObject.cpp:1421]: I think RegExpPrototype watchpoint optimization is probably missing. Please grep all use of existing getters carefully and align the new one with this.

Comment (msaboff) [reply Source/JavaScriptCore/runtime/JSGlobalObject.cpp:1421]: I grep'ed the other uses of the RegExp getters and only find uses in DFGAbstractInterpreterInlines.h and DFGFixupPhase.  The unicodeSets getter is already added to those sections.

**Review by Constellation — APPROVED** [review]
rs=me with watchpoint fix.

File: Source/JavaScriptCore/runtime/JSGlobalObject.cpp (hunk 5 above)
Comment (msaboff) [reply Source/JavaScriptCore/runtime/JSGlobalObject.cpp:1421]: Fixed.

**Pull request merged**